packages/snip20/src/query.rs

```diff
@@ -0,0 +1,315 @@
+use core::fmt;
+use schemars::JsonSchema;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+use cosmwasm_std::{
+    to_binary, Api, Coin, Extern, HumanAddr, Querier, QueryRequest, StdError, StdResult, Storage,
+    Uint128, WasmQuery,
+};
+
+use secret_toolkit_utils::space_pad;
+
+/// TokenInfo response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct TokenInfo {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u8,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub total_supply: Option<Uint128>,
+}
+/// ExchangeRate response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct ExchangeRate {
+    pub rate: Uint128,
+    pub denom: String,
+}
+/// Allowance response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Allowance {
+    pub spender: HumanAddr,
+    pub owner: HumanAddr,
+    pub allowance: Uint128,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub expiration: Option<u64>,
+}
+/// Balance response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Balance {
+    pub amount: Uint128,
+}
+/// Transaction data
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Tx {
+    pub id: u64,
+    pub from: HumanAddr,
+    pub sender: HumanAddr,
+    pub receiver: HumanAddr,
+    pub coins: Coin,
+}
+/// TransferHistory response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct TransferHistory {
+    pub txs: Vec<Tx>,
+}
+/// Minters response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Minters {
+    pub minters: Vec<HumanAddr>,
+}
+
+/// SNIP20 queries
+#[derive(Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Snip20QueryMsg<'a> {
+    TokenInfo {},
+    ExchangeRate {},
+    Allowance {
+        owner: &'a HumanAddr,
+        spender: &'a HumanAddr,
+        key: &'a str,
+    },
+    Balance {
+        address: &'a HumanAddr,
+        key: &'a str,
+    },
+    TransferHistory {
+        address: &'a HumanAddr,
+        key: &'a str,
+        page: Option<u32>,
+        page_size: u32,
+    },
+    Minters {},
+}
+
+impl<'a> fmt::Display for Snip20QueryMsg<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Snip20QueryMsg::TokenInfo { .. } => write!(f, "TokenInfo"),
+            Snip20QueryMsg::ExchangeRate { .. } => write!(f, "ExchangeRate"),
+            Snip20QueryMsg::Allowance { .. } => write!(f, "Allowance"),
+            Snip20QueryMsg::Balance { .. } => write!(f, "Balance"),
+            Snip20QueryMsg::TransferHistory { .. } => write!(f, "TransferHistory"),
+            Snip20QueryMsg::Minters { .. } => write!(f, "Minters"),
+        }
+    }
+}
```

Comment:
what do you need this for?

Reply:
I was thinking it would be nice to "wrap" an error returned from doing the "query" function with:
```rust
            .map_err(|err| {
                StdError::generic_err(format!("Error performing {} query: {}", self, err))
            })
```
So if someone was doing multiple queries and they got some generic Deserialization error, they would know which query it came from because it would be prefaced with "Error performing XXX query:"

Reply:
ahhhh yes this makes sense. I was thinking that `#[derive(Debug)]` is all you need, but this lets users get just the variant name, which is nice :)